## This Script uses DESeq2 to identify differentially expressed genes between tumour groups output from the RNA-seq-2022 dataset. ##
## Variables to set include the raw counts file output obtained from the Featurecounts software (rawdata), an information matrix about the differet experimental
## conditions of the collumns present on the raw featurecounts output file. Experimental group acts as a suffix for output files generated by the Deseq2 package
## Group1 and Group2 variables are prefixes used to generate DESeq2 output files and should be defined as the experimental groups/collumns the user wishes to run
##in DESeq2

###LIBRARIES###
library(ggplot2)
library(reshape2)
library(DESeq2)
library(pheatmap)
library(dplyr)

###VARIABLES TO SET###
rawdata <- read.csv("<RAWCOUNTS_FILE>", header = T, row.names = 1) # <- Rawcounts Matrix File Name.
info <- read.csv("<INFO_FILE>", header = T, row.names = 1) # <- Info Matrix File Name.
experimentalGroup <- "<EXPERIMENT_NAME>" # <- What does the rawcounts file contain? - eg. "All_WD_samples"
experimentalCondition <- "<EXPERIMENTAL_CONDITION>" # <- Which experimental condition to test on <INFO_FILE> eg. GenotypeDay
group1 <- "<GENOTYPE_1>" # 1st Genotype of Rawcounts matrix file for the 2 way comparison
group2 <- "<GENOTYPE_2>" # 2nd Genotype of Rawcounts matrix file for the 2 way comparison
lowCountThreshold <- 50 # 50 Recommended.
minLogFoldChange <- 1.4 #1.4 Recommended.
minPvalue <- 0.05 #0.05 Recommended.

###FILENAME GENERATOR###
normalisedMatrixOutputFilename <- paste0(groupA,"_V_",groupB,"_Normalised_Counts.csv") # <- May need to specify output file name for normalised counts matrix.
deseqResultsFilename <- paste0(groupA,"_V_",groupB,"_DESEQ_Results.csv")
sigFoldUpFilename  <- paste0(groupA,"_V_",groupB,"_Significant_Upregulated_Genes.csv")
sigFoldDownFilename <- paste0(groupA,"_V_",groupB,"_Significant_Downregulated_Genes.csv")
allSigGenesFilename <- paste0(groupA,"_V_",groupB,"_All_Significant_Genes.csv")

###RUN DESEQ2###
#Create DESeq data set using the two reference files outlined above.
dds <- DESeqDataSetFromMatrix(rawdata, info, ~GenotypeDay)
#Remove lowly expressed genes from the data set that have a total row count of below.
keep <- rowSums(counts(dds)) >= lowCountThreshold
dds <- dds[keep,]
#Run Deseq normalization. Factors in difference in library sizes and estimates dispersion levels.
ddsDE <- DESeq(dds)
#Extract All Normalised Counts.
normcounts <- counts(ddsDE, normalized = T )
#Export normalized read counts to a .cvs file for later use.
write.csv(normcounts, normalisedMatrixOutputFilename)
#Run DESEQ2 on normalized counts file.
resultsNames(ddsDE)
res <- results(ddsDE, contrast=c("GenotypeDay", groupB, groupA))
statsDE <- res[order(res$padj),]
write.csv(statsDE, deseqResultsFilename)

###DIFFERENTIAL GENE IDENTIFIER### (For 2 Variable Analysis)###
#Remove all genes with a total row count of < X.
moreThanFifty <- subset(statsDE, baseMean > lowCountThreshold)
#Remove all genes with a padj (ajusted pvalue) of less that X.
filteredStatsDE <- subset(moreThanFifty, padj < minPvalue)
#Pull Out all genes with a log fold change of > X into a list and export to CSV
sigFoldGenesUp <- subset(filteredStatsDE, log2FoldChange > minLogFoldChange)
write.csv(sigFoldGenesUp, sigFoldUpFilename)
#Pull out all genes with a log fold change of < X into a list and export to a CSV
sigFoldGenesDown <- subset(filteredStatsDE, log2FoldChange < -minLogFoldChange)
write.csv(sigFoldGenesDown, sigFoldDownFilename)
#Write all Differentially expressed genes to a csv file#
sigFoldAllGenes <- rbind(sigFoldGenesUp, sigFoldGenesDown)
write.csv(sigFoldAllGenes, allSigGenesFilename)
